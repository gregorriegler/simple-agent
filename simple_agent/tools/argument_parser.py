import shlex


def create_lexer(text):
    lexer = shlex.shlex(text, posix=True)
    lexer.whitespace_split = True
    lexer.commenters = ''
    lexer.escape = ''
    return lexer


def split_arguments(text):
    lexer = create_lexer(text)
    return list(lexer)
