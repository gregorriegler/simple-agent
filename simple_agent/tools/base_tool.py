import subprocess
from simple_agent.application.tool_library import ToolResult


class BaseTool:
    name = ''
    description = ''
    arguments = []

    def execute(self, *args, **kwargs) -> ToolResult:
        raise NotImplementedError("Subclasses must implement execute()")

    @staticmethod
    def run_command(command, args=None, cwd=None):
        try:
            command_line = [command]
            if args:
                if isinstance(args, str):
                    args = [args]
                command_line += args

            result = subprocess.run(
                command_line,
                stdin=subprocess.DEVNULL,
                stdout=subprocess.PIPE,
                stderr=subprocess.PIPE,
                text=True,
                encoding='utf-8',
                errors='replace',
                timeout=30,
                cwd=cwd
            )
            output = result.stdout.rstrip('\n')
            if result.stderr:
                if output:
                    output += f"\n";
                output += f"STDERR: {result.stderr}"
            return {
                'output': output
            }
        except subprocess.TimeoutExpired:
            return {'output': 'Command timed out (30s limit)'}
        except Exception as e:
            return {'output': f'Error: {str(e)}'}

    def get_usage_info(self):
        if hasattr(self, '_custom_usage_info'):
            return self._custom_usage_info()

        return self._generate_usage_info_from_metadata()

    def _generate_usage_info_from_metadata(self):
        lines = [f"Tool: {self.name}"]

        if self.description:
            lines.append(f"Description: {self.description}")

        if self.arguments:
            lines.append("")
            lines.append("Arguments:")
            for arg in self.arguments:
                required_str = " (required)" if arg.get('required', False) else " (optional)"
                type_str = f" - {arg['name']}: {arg.get('type', 'string')}{required_str}"
                if 'description' in arg:
                    type_str += f" - {arg['description']}"
                lines.append(type_str)

            # Generate syntax
            lines.append("")
            required_args = [f"<{arg['name']}>" for arg in self.arguments if arg.get('required', False)]
            optional_args = [f"[{arg['name']}]" for arg in self.arguments if not arg.get('required', False)]
            all_args = required_args + optional_args
            syntax = f"üõ†Ô∏è {self.name}"
            if all_args:
                syntax += " " + " ".join(all_args)
            lines.append(f"Usage: {syntax}")

        if hasattr(self, 'examples') and self.examples:
            lines.append("")
            lines.append("Examples:")
            for example in self.examples:
                lines.append(f"{example}")

        return "\n".join(lines)
