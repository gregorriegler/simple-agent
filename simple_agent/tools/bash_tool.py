import subprocess
from ..application.tool_library import ContinueResult
from .base_tool import BaseTool

class BashTool(BaseTool):
    name = 'bash'
    description = "Execute bash commands"
    arguments = [
        {
            "name": "command",
            "type": "string",
            "required": True,
            "description": "The bash command to execute"
        }
    ]
    examples = [
        "🛠️ bash echo hello",
        "🛠️ bash ls -la",
        "🛠️ bash pwd"
    ]

    def execute(self, args):
        if not args:
            return ContinueResult('STDERR: bash: missing command', success=False)
        _ = subprocess
        result = self.run_command("bash", ["-c", args])
        return ContinueResult(result['output'], success=result['success'])
