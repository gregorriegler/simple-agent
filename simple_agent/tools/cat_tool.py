from ..application.tool_library import ContinueResult
from .base_tool import BaseTool
from .argument_parser import split_arguments


class CatTool(BaseTool):
    name = 'cat'
    description = "Display file contents with line numbers"
    arguments = [
        {
            "name": "filename",
            "type": "string",
            "required": True,
            "description": "Path to the file to display"
        },
        {
            "name": "line_range",
            "type": "string",
            "required": False,
            "description": "Optional line range in format 'start-end' (e.g., '1-10')"
        }
    ]
    examples = [
        "🛠️ cat myfile.txt",
        "🛠️ cat script.py 1-20"
    ]

    def _parse_arguments(self, args):
        if not args:
            return None, None, 'STDERR: cat: missing file operand'

        try:
            parts = split_arguments(args)
        except ValueError as exc:
            return None, None, f"STDERR: cat: {exc}"

        if not parts:
            return None, None, 'STDERR: cat: missing file operand'

        filename = parts[0]
        line_range = parts[1] if len(parts) > 1 else None

        return filename, line_range, None

    def _validate_range(self, line_range):
        try:
            start_line, end_line = map(int, line_range.split('-'))
        except ValueError:
            return None, None, f"STDERR: Invalid range format '{line_range}'. Use format 'start-end' (e.g., '1-5')"

        if start_line > end_line:
            return None, None, f"STDERR: Start line ({start_line}) cannot be greater than end line ({end_line})"

        return start_line, end_line, None

    def execute(self, args):
        filename, line_range, error = self._parse_arguments(args)
        if error:
            return ContinueResult(error)
        if line_range is None:
            result = self.run_command('cat', ['-n', filename])
            return ContinueResult(result['output'])
        start_line, end_line, error = self._validate_range(line_range)
        if error:
            return ContinueResult(error)
        output = self._read_file_range(filename, start_line, end_line)
        return ContinueResult(output)

    def _read_file_range(self, filename, start_line, end_line):
        try:
            with open(filename, 'r', encoding='utf-8') as f:
                lines = f.readlines()

            if not lines:
                return ""

            start_idx = start_line - 1
            end_idx = min(end_line, len(lines))

            if start_idx >= len(lines):
                return ""

            return self._format_output(lines, start_idx, end_idx)

        except FileNotFoundError:
            return f"STDERR: cat: '{filename}': No such file or directory"
        except Exception as e:
            return f"STDERR: cat: '{filename}': {str(e)}"

    def _format_output(self, lines, start_idx, end_idx):
        result_lines = []
        for i in range(start_idx, end_idx):
            result_lines.append(f"{i + 1:6}\t{lines[i]}")

        return "".join(result_lines).rstrip('\n')
