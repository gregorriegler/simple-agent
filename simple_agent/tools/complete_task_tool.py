from ..application.tool_library import CompleteResult
from .base_tool import BaseTool
class CompleteTaskTool(BaseTool):
    name = "complete-task"
    description = "Signal task completion with a summary of what was accomplished"
    arguments = [
        {
            "name": "summary",
            "type": "string",
            "required": True,
            "description": "Final summary of what was accomplished"
        }
    ]
    examples = [
        "üõ†Ô∏è complete-task Successfully created the user registration system",
        "üõ†Ô∏è complete-task Fixed the bug in the payment processing module"
    ]

    def execute(self, args):
        if not args or not args.strip():
            return CompleteResult('STDERR: complete-task: missing summary')
        summary = args.strip()
        return CompleteResult(summary)

