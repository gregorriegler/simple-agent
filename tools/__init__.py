from .base_tool import BaseTool
from .tool_library import ToolLibrary, ParsedTool

__all__ = [
    'BaseTool',
    'ToolLibrary',
    'ParsedTool'
]
