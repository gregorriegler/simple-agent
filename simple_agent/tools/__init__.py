from .base_tool import BaseTool
from .all_tools import AllTools, ParsedTool

__all__ = [
    'BaseTool',
    'AllTools',
    'ParsedTool'
]
