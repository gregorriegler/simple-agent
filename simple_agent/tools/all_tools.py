import re
from typing import Callable, Any

from simple_agent.application.agent import Agent
from simple_agent.application.event_bus_protocol import EventBus
from simple_agent.application.input import Input
from simple_agent.application.llm import LLM
from simple_agent.application.session_storage import NoOpSessionStorage
from simple_agent.application.tool_library import ToolLibrary, MessageAndParsedTools, ParsedTool, Tool
from simple_agent.system_prompt_generator import generate_system_prompt
from .bash_tool import BashTool
from .cat_tool import CatTool
from .complete_task_tool import CompleteTaskTool
from .create_file_tool import CreateFileTool
from .edit_file_tool import EditFileTool
from .ls_tool import LsTool
from .subagent_tool import SubagentTool
from .write_todos_tool import WriteTodosTool


class AllTools(ToolLibrary):
    def __init__(
        self,
        llm: LLM | None = None,
        indent_level=0,
        agent_id: str = "Agent",
        event_bus: EventBus | None = None,
        user_input=None,
        create_subagent_display: Callable[[str, int], Any] | None = None,
        create_subagent_input: Callable[[int], Input] | None = None
    ):
        from simple_agent.application.event_bus import SimpleEventBus

        self.llm: LLM = llm if llm is not None else (lambda system_prompt, messages: '')
        self.indent_level = indent_level
        self.agent_id = agent_id
        self.event_bus: EventBus = event_bus if event_bus is not None else SimpleEventBus()
        self.user_input = user_input

        assert create_subagent_display is not None, "create_subagent_display must be provided"
        assert create_subagent_input is not None, "create_subagent_input must be provided"

        self.create_subagent_display = create_subagent_display
        self.create_subagent_input = create_subagent_input

        static_tools = self._create_static_tools()
        dynamic_tools = self._discover_dynamic_tools()
        self.tools: list[Tool] = static_tools + dynamic_tools + [self._create_subagent_tool()]
        self.tool_dict = {getattr(tool, 'name', ''): tool for tool in self.tools}

    def _create_static_tools(self):
        return [
            WriteTodosTool(self.agent_id),
            LsTool(),
            CatTool(),
            CreateFileTool(),
            EditFileTool(),
            #            PatchFileTool(),
            #            RememberTool(),
            #           RecallTool(),
            CompleteTaskTool(),
            BashTool()
        ]

    def _discover_dynamic_tools(self):
        return []

    def _create_subagent_tool(self):
        return SubagentTool(
            self._create_default_agent,
            self.create_subagent_display,
            self.indent_level,
            self.agent_id,
            self.create_subagent_input
        )

    def _create_default_agent(
        self,
        agent_id,
        user_input,
        session_storage=NoOpSessionStorage()
    ) -> Agent:
        subagent_tools = AllTools(
            self.llm,
            self.indent_level + 1,
            agent_id,
            self.event_bus,
            self.user_input,
            self.create_subagent_display,
            self.create_subagent_input
        )
        system_prompt = self._build_system_prompt(subagent_tools)
        return Agent(
            agent_id,
            system_prompt,
            subagent_tools,
            self.llm,
            user_input,
            self.event_bus,
            session_storage
        )

    def _build_system_prompt(self, subagent_tools) -> Callable[[Any], str]:
        return lambda tool_library: generate_system_prompt(subagent_tools)

    def parse_message_and_tools(self, text) -> MessageAndParsedTools:
        pattern = r'^🛠️ ([\w-]+)(?:\s+(.*))?'
        end_marker = r'^🛠️🔚'
        lines = text.splitlines(keepends=True)
        parsed_tools = []
        message = ""
        first_tool_index = None

        i = 0
        while i < len(lines):
            match = re.match(pattern, lines[i], re.DOTALL)
            if match:
                if first_tool_index is None:
                    first_tool_index = i
                    message = ''.join(lines[:i]).rstrip()

                command, same_line_args = match.groups()
                tool = self.tool_dict.get(command)
                if not tool:
                    return MessageAndParsedTools(message=text, tools=[])

                all_arg_lines = []
                if same_line_args:
                    all_arg_lines.append(same_line_args)

                i += 1
                while i < len(lines) and not re.match(r'^🛠️ ', lines[i]) and not re.match(end_marker, lines[i]):
                    all_arg_lines.append(lines[i])
                    i += 1

                if i < len(lines) and re.match(end_marker, lines[i]):
                    i += 1

                arguments = ''.join(all_arg_lines).rstrip()
                parsed_tools.append(ParsedTool(command, arguments, tool))
            else:
                i += 1

        if parsed_tools:
            return MessageAndParsedTools(message=message, tools=parsed_tools)
        return MessageAndParsedTools(message=text, tools=[])

    def execute_parsed_tool(self, parsed_tool):
        args = parsed_tool.arguments if parsed_tool.arguments else None
        result = parsed_tool.tool_instance.execute(args)
        return result
