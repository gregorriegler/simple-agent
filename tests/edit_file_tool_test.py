import os
from approvaltests import Options, verify

from simple_agent.tools.all_tools import AllTools
from .test_helpers import all_scrubbers, temp_directory

library = AllTools()


def test_edit_file_replace_single_word(tmp_path):
    verify_edit_tool(library, "test.txt", "old", "üõ†Ô∏è edit-file test.txt replace 1-1\nnew", tmp_path=tmp_path)


def test_edit_file_replace_two_consecutive_lines_with_one(tmp_path):
    verify_edit_tool(library, "test.txt", "line1\nline2\nline3", "üõ†Ô∏è edit-file test.txt replace 2-3\nnewline", tmp_path=tmp_path)


def test_edit_file_replace_a_lines_with_lines_in_quotes(tmp_path):
    command = "üõ†Ô∏è edit-file test.txt replace 2\n\"insert1\ninsert2\""
    verify_edit_tool(library, "test.txt", "line1\nline2", command, tmp_path=tmp_path)


def test_edit_file_replace_empty_lines_with_function(tmp_path):
    initial_content = "line1\nline2\n\n\n\nline6"
    function_def = "def hello():\n    return 'world'"
    command = f"üõ†Ô∏è edit-file template.py replace 3-5\n{function_def}"
    verify_edit_tool(library, "template.py", initial_content, command, tmp_path=tmp_path)


def test_edit_file_replace_path_with_spaces(tmp_path):
    command = 'üõ†Ô∏è edit-file "notes folder/note file.txt" replace 1\n"updated line"'
    verify_edit_tool(library, "notes folder/note file.txt", "original line", command, tmp_path=tmp_path)


def test_edit_file_insert_three_lines_to_empty_file(tmp_path):
    command = "üõ†Ô∏è edit-file empty.txt insert 1\nline1\nline2\nline3"
    verify_edit_tool(library, "empty.txt", "", command, tmp_path=tmp_path)


def test_edit_file_insert_line_without_explicit_newline_adds_newline_automatically(tmp_path):
    initial_content = "line1\nline2"
    command = "üõ†Ô∏è edit-file test.txt insert 2\ninserted_line"
    verify_edit_tool(library, "test.txt", initial_content, command, tmp_path=tmp_path)


def test_edit_file_insert_multiline_without_trailing_newline_adds_newline_automatically(tmp_path):
    initial_content = "line1\nline2"
    command = "üõ†Ô∏è edit-file test.txt insert 2\ninserted_line1\ninserted_line2"
    verify_edit_tool(library, "test.txt", initial_content, command, tmp_path=tmp_path)


def test_edit_file_preserves_indentation_in_multiline_content(tmp_path):
    initial_content = "line1\nline2"
    command = "üõ†Ô∏è edit-file test.py insert 3\ndef hello():\n    return 'world'"
    verify_edit_tool(library, "test.py", initial_content, command, tmp_path=tmp_path)


def test_edit_file_insert_beyond_last_line_appends_to_end(tmp_path):
    initial_content = "line1\nline2"
    command = "üõ†Ô∏è edit-file test.txt insert 3\nline3"
    verify_edit_tool(library, "test.txt", initial_content, command, tmp_path=tmp_path)


def test_edit_file_insert_far_beyond_last_line_appends_to_end_and_pads_with_empty_lines(tmp_path):
    initial_content = "line1\nline2"
    command = "üõ†Ô∏è edit-file test.txt insert 10\nline10"
    verify_edit_tool(library, "test.txt", initial_content, command, tmp_path=tmp_path)


def test_edit_file_insert_without_content_inserts_an_empty_line(tmp_path):
    initial_content = "line1\nline2"
    command = "üõ†Ô∏è edit-file test.txt insert 2"
    verify_edit_tool(library, "test.txt", initial_content, command, tmp_path=tmp_path)


def test_edit_file_delete_a_line(tmp_path):
    initial_content = "line1\nline2"
    command = "üõ†Ô∏è edit-file test.txt delete 1"
    verify_edit_tool(library, "test.txt", initial_content, command, tmp_path=tmp_path)


def test_edit_file_delete_range_extending_past_file_end(tmp_path):
    initial_content = "line1\nline2\nline3"
    command = "üõ†Ô∏è edit-file test.txt delete 2-10"
    verify_edit_tool(library, "test.txt", initial_content, command, tmp_path=tmp_path)


def test_edit_file_replace_range_beyond_file_end_leaves_file_unchanged(tmp_path):
    initial_content = "line1\nline2"
    command = "üõ†Ô∏è edit-file test.txt replace 5-7\nreplacement"
    verify_edit_tool(library, "test.txt", initial_content, command, tmp_path=tmp_path)


def test_edit_file_replace_with_auto_indent_python(tmp_path):
    initial_content = "line1\n    existing = 1\nline3"
    command = "üõ†Ô∏è edit-file test.py replace 2\nnew = 2"
    verify_edit_tool(library, "test.py", initial_content, command, tmp_path=tmp_path)


def test_edit_file_insert_with_auto_indent(tmp_path):
    initial_content = "line1\n    line2\nline3"
    command = "üõ†Ô∏è edit-file test.py insert 2\nnew_line"
    verify_edit_tool(library, "test.py", initial_content, command, tmp_path=tmp_path)


def test_edit_file_replace_preserves_manual_indentation(tmp_path):
    initial_content = "line1\n    existing\nline3"
    command = "üõ†Ô∏è edit-file test.py replace 2\n        manually_indented"
    verify_edit_tool(library, "test.py", initial_content, command, tmp_path=tmp_path)


def test_edit_file_replace_multiline_only_indents_first_line(tmp_path):
    initial_content = "line1\n    existing\nline3"
    command = "üõ†Ô∏è edit-file test.py replace 2\nline1\nline2\n    line3"
    verify_edit_tool(library, "test.py", initial_content, command, tmp_path=tmp_path)


def test_edit_file_replace_complex_python_program_rename_functions_and_variables(tmp_path):
    initial_content = """class DataProcessor:
    def __init__(self, input_data):
        self.raw_data = input_data
        self.processed_items = []

    def validate_input(self, data_item):
        if not data_item:
            return False
        return len(data_item) > 0

    def process_single_item(self, item):
        if self.validate_input(item):
            cleaned_item = item.strip().lower()
            return cleaned_item
        return None

    def batch_process(self):
        for current_item in self.raw_data:
            result = self.process_single_item(current_item)
            if result:
                self.processed_items.append(result)
        return self.processed_items"""

    replacement_content = """class DataHandler:
    def __init__(self, source_data):
        self.original_data = source_data
        self.cleaned_items = []

    def check_validity(self, data_entry):
        if not data_entry:
            return False
        return len(data_entry) > 0

    def handle_single_entry(self, entry):
        if self.check_validity(entry):
            normalized_entry = entry.strip().lower()
            return normalized_entry
        return None

    def process_all(self):
        for current_entry in self.original_data:
            outcome = self.handle_single_entry(current_entry)
            if outcome:
                self.cleaned_items.append(outcome)
        return self.cleaned_items"""

    command = f"üõ†Ô∏è edit-file complex_program.py replace 1-20\n{replacement_content}"
    verify_edit_tool(library, "complex_program.py", initial_content, command, tmp_path=tmp_path)


def verify_edit_tool(library, setup_file, setup_content, command, tmp_path):
    with temp_directory(tmp_path):
        os.makedirs(os.path.dirname(setup_file) or '.', exist_ok=True)
        with open(setup_file, "w", encoding='utf-8') as f:
            f.write(setup_content)

        initial_file_info = f"Initial file: {setup_file}\nInitial content:\n--- INITIAL CONTENT START ---\n{setup_content}\n--- INITIAL CONTENT END ---"

        tool = library.parse_tool(command)
        result = library.execute_parsed_tool(tool.tools[0])
        with open(setup_file, "r", encoding='utf-8') as f:
            actual_content = f.read()
        final_file_info = f"File after edit: {setup_file}\nFinal content:\n--- FINAL CONTENT START ---\n{actual_content}\n--- FINAL CONTENT END ---"
        verify(
            f"Command:\n{command}\n\nResult:\n{result}\n\n{initial_file_info}\n\n{final_file_info}",
            options=Options().with_scrubber(all_scrubbers())
        )
