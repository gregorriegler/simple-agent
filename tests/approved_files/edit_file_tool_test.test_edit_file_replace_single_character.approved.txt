Command:
üõ†Ô∏è edit-file test.txt replace 1-1 b

Result:
Successfully edited test.txt, lines 1-1

Initial file: test.txt
Initial content:
--- INITIAL CONTENT START ---
a
--- INITIAL CONTENT END ---

File after edit: test.txt
Final content:
--- FINAL CONTENT START ---
b
--- FINAL CONTENT END ---
